File: 17_letter_combination_of_a_phone_number.py
def letterCombinations(self, digits: str) -> List[str]:
        
        self.dig2l = {'2': ['a', 'b', 'c'], '3': ['d', 'e', 'f'], '4':['g', 'h', 'i'], '5': ['j', 'k', 'l'], '6': ['m', 'n', 'o'], '7': ['p', 'q', 'r', 's'], '8': ['t', 'u', 'v'], '9': ['w', 'x', 'y', 'z']}
        
        res, self.dig_length = [], len(digits)
  
        self.build_combos(res, digits)
        return res
        
    def build_combos(self, res, digits:str, combination = ''):
        
        if len(combination) == self.dig_length:
            if combination: res.append(combination)
            return
        
        for char in self.dig2l[digits[0]]:
            self.build_combos(res, digits[1:],combination + char)

File: 17_letter_combination_of_a_phone_number.cpp
class Solution {
public:
    
    void dfs(vector<string> &res, string str,string digits,int l,unordered_map<char,vector<char>>m){
        if(str.length()==digits.length()){
            res.push_back(str);
            return;
        }
        char c = digits[l];
        for(auto j=m[c].begin();j!=m[c].end();j++){
            dfs(res,str+(*j),digits,l+1,m);
        }
    }
    
    vector<string> letterCombinations(string digits) {
        unordered_map<char,vector<char>>m;
        m['2'].insert(m['2'].begin(),{'a','b','c'});
        m['3'].insert(m['3'].begin(),{'d','e','f'});
        m['4'].insert(m['4'].begin(),{'g','h','i'});
        m['5'].insert(m['5'].begin(),{'j','k','l'});
        m['6'].insert(m['6'].begin(),{'m','n','o'});
        m['7'].insert(m['7'].begin(),{'p','q','r','s'});
        m['8'].insert(m['8'].begin(),{'t','u','v'});
        m['9'].insert(m['9'].begin(),{'w','x','y','z'});
        
        // for(auto i=m.begin();i!=m.end();i++){
        //     for(int j=0;j<i->second.size();j++){
        //         cout<<i->second[j]<<" ";
        //     }
        //     cout<<endl;
        // }
        vector<string> res;
        if(!digits.size())
            return res;
        dfs(res,"",digits,0,m);
        return res;
    }
};